core/src/main/java/org/elasticsearch/indices/IndicesModule.java

```diff
@@ -138,6 +221,23 @@ protected void configure() {
         bind(NodeServicesProvider.class).asEagerSingleton();
     }
 
+    // public for testing
+    public synchronized MapperRegistry getMapperRegistry() {
+        // NOTE: we register _field_names here so that it has a chance to see all other
+        // mappers, including from plugins
+        if (metadataMapperParsers.containsKey(FieldNamesFieldMapper.NAME)) {
+            throw new IllegalStateException("Metadata mapper [" + FieldNamesFieldMapper.NAME + "] is already registered");
+        }
+        final Map<String, MetadataFieldMapper.TypeParser> metadataMapperParsers
+            = new LinkedHashMap<>(this.metadataMapperParsers);
+        metadataMapperParsers.put(FieldNamesFieldMapper.NAME, new FieldNamesFieldMapper.TypeParser());
+        return new MapperRegistry(mapperParsers, metadataMapperParsers);
```

Comment:
I'd have added an integer to `TypeParser` and sorted them by the integer resolving same integers alphabetically or something. And set the FieldNamesFieldMapper to MAXINT. But I think what you did here is ultimately simpler.